### ISSUE TITLE
Overlapping overload with mismatched return not detected for parent/child class

### ISSUE THREAD

**Issue by jakebailey** [microsoft/pyright#1572]
The following code fails in mypy due to an overlapping overload, but does not fail in pyright.

```python
from typing import Union, overload

class Parent: ...
class Child(Parent): ...

@overload
def func(x: Child) -> int: ...
@overload
def func(x: Parent) -> str: ...
def func(x: Union[Child, Parent]) -> Union[int, str]:
    if isinstance(x, Child):
        return 1234
    return "hey"
```

```
main.py:7: error: Overloaded function signatures 1 and 2 overlap with incompatible return types
Found 1 error in 1 file (checked 1 source file)
```

This shouldn't be legal because the result because the result here would be wrong:

```python
def get_parent() -> Parent:
    return Child()

x: Parent = get_parent()
v = func(x)  # v is statically str, but at runtime is int.
```

**Comment by erictraut** [comment]
I'm not able to repro. For me, pyright properly detects and reports this error. Did you forget to enable the `reportOverlappingOverload` diagnostic check (or enable strict checking)?

<img width="799" alt="Screen Shot 2021-03-04 at 3 05 39 PM" src="https://user-images.githubusercontent.com/7040122/110037026-17087d80-7cfb-11eb-8808-a9271e308306.png">

**Comment by jakebailey** [comment]
I have strict mode enabled. This is in the stable Pylance build. Note the slew of things here:

![image](https://user-images.githubusercontent.com/5341706/110037293-10760800-7cf3-11eb-93b2-042fb95536d3.png)

If I manually enable the check, then it works. Now that I look, `reportOverlappingOverload` isn't actually enabled in strict mode; it's set to `none` for each preset. I guess I thought it would be enabled in strict mode (or even basic mode).

**Comment by jakebailey** [comment]
Seems like the schema and package.json are all consistent about the default of this being `reportOverlappingOverload`, so I guess if it's intended that it not be enabled in strict/basic, then I can close this.

**Comment by erictraut** [comment]
Oh, that's right. I removed that from the strict rule set. You have to enable it explicitly even in strict mode. We could reverse that decision and add it back to "strict". It needs to remain off in the typeshed CI tests though because there are five violations it finds. Mypy has two bugs that prevents it from catching some violations. Both of these bugs are filed against mypy already.

**Comment by jakebailey** [comment]
Up to you; the config in typeshed disables it explicitly, so I would be for adding it back to strict if we believe it to be important to catch. If mypy emits errors for this category in its default mode, then that'd be evidence it might be important for basic, but I would rather not experiment too heavily with that (and I don't know all of the little differences between basic and mypy).

**Comment by erictraut** [comment]
I think it makes sense to enable this check in strict mode. I'm uncomfortable doing so in basic mode at this time.

**Comment by erictraut** [comment]
This is changed in Pyright 1.1.118, which I just published. It will also be included in the next release of Pylance.
